리뷰, 후기 기능 #1

shushuaka opened the pull request:
### #1 리뷰, 후기  관리 기능을 구현했습니다~

1. **엔티티 및 리포지토리 추가**
   - `Review.kt`: 리뷰 데이터를 저장하는 JPA 엔티티 추가
   - `ReviewRepository.kt`: 데이터베이스 접근 및 조회 메서드(`findByUserIdAndShowId`, `findAllByShowId`) 추가

2. **DTO 추가**
   - `ReviewReqDto.kt`: 리뷰 작성 및 수정 요청 데이터 처리
   - `ReviewResDto.kt`: 리뷰 조회 및 응답 데이터 반환

3. **비즈니스 로직 구현**
   - `ReviewService.kt`: 리뷰 생성, 조회, 수정, 삭제 로직 구현
   - 예외 처리 및 데이터 일관성을 보장하기 위해 `@Transactional` 적용

4. **컨트롤러 구현**
   - `ReviewController.kt`: 리뷰 조회, 생성, 수정, 삭제 API 엔드포인트 추가
   - 서비스 호출을 통해 비즈니스 로직 위임

5. **매퍼 추가**
   - `ReviewMapper.kt`: 엔티티와 DTO 간 변환 로직 구현

6. **예외 처리 확장**
   - `CustomException.kt`, `ExceptionResponse.kt`, `ExceptionResponseStatus.kt`:
     - 리뷰 관련 예외 상태 추가 (`REVIEW_NOT_FOUND`, `REVIEW_ALREADY_EXISTS` 등)
   - 예외 처리를 위한 `GlobalExceptionHandler.kt` 활용

---

### **주요 기능**
- 리뷰 조회: `GET /api/reviews` (옵션으로 `showId` 필터링 가능)
- 리뷰 생성: `POST /api/reviews`
- 리뷰 수정: `PATCH /api/reviews/{reviewId}`
- 리뷰 삭제: `DELETE /api/reviews/{reviewId}`

---

### **확인 요청**
- 모든 API가 올바르게 동작하는지 확인 부탁드립니다. (저는 잘 됨)
- 코드 스타일과 예외 처리 방식에 대한 피드백을 요청드립니다.

감사합니다! 🐧

shushuaka committed:
feat: 리뷰 엔티티 및 리포지토리 추가
- 리뷰 데이터를 관리하기 위한 JPA 엔티티(Review) 생성
- 리뷰 엔티티와 데이터베이스를 연동하기 위한 리포지토리(ReviewRepository) 구현

shushuaka committed:
feat: 리뷰 엔티티 및 리포지토리 추가
- 리뷰 데이터를 관리하기 위한 JPA 엔티티(Review) 생성
- 리뷰 엔티티와 데이터베이스를 연동하기 위한 리포지토리(ReviewRepository) 구현

shushuaka committed:
feat: 리뷰 요청/응답 DTO 추가
- 클라이언트에서 리뷰 데이터를 전달받기 위한 ReviewReqDto 추가
- 클라이언트로 리뷰 데이터를 반환하기 위한 ReviewResDto 추가
- 유효성 검증(@NotBlank, @Size, @Min, @Max)을 ReviewReqDto에 적용

shushuaka committed:
feat: 리뷰 요청/응답 DTO 추가
- 클라이언트에서 리뷰 데이터를 전달받기 위한 ReviewReqDto 추가
- 클라이언트로 리뷰 데이터를 반환하기 위한 ReviewResDto 추가
- 유효성 검증(@NotBlank, @Size, @Min, @Max)을 ReviewReqDto에 적용

shushuaka committed:
feat: 리뷰 관리 (생성, 조회, 수정, 삭제) 유스케이스 구현
- 리뷰 생성(CreateReviewUseCase): 사용자가 새로운 리뷰를 작성할 수 있도록 구현
- 리뷰 조회(GetReviewsUseCase): 특정 쇼 또는 모든 리뷰를 조회할 수 있도록 구현
- 리뷰 수정(UpdateReviewUseCase): 기존 리뷰의 댓글과 평점을 수정할 수 있도록 구현
- 리뷰 삭제(DeleteReviewUseCase): 특정 리뷰를 삭제할 수 있도록 구현
- 각 유스케이스에 CustomException을 활용한 예외 처리 추가

shushuaka committed:
feat: 리뷰 관리 (생성, 조회, 수정, 삭제) 유스케이스 구현
- 리뷰 생성(CreateReviewUseCase): 사용자가 새로운 리뷰를 작성할 수 있도록 구현
- 리뷰 조회(GetReviewsUseCase): 특정 쇼 또는 모든 리뷰를 조회할 수 있도록 구현
- 리뷰 수정(UpdateReviewUseCase): 기존 리뷰의 댓글과 평점을 수정할 수 있도록 구현
- 리뷰 삭제(DeleteReviewUseCase): 특정 리뷰를 삭제할 수 있도록 구현
- 각 유스케이스에 CustomException을 활용한 예외 처리 추가

shushuaka committed:
feat: 리뷰 엔티티와 DTO 매핑 기능 추가
- Review 엔티티와 ReviewReqDto/ReviewResDto 간 변환 기능 구현
- 리뷰 생성 및 조회 시 데이터 변환을 위해 ReviewMapper 활용
- 엔티티와 DTO를 분리하여 유지보수성을 향상

shushuaka committed:
feat: 리뷰 엔티티와 DTO 매핑 기능 추가
- Review 엔티티와 ReviewReqDto/ReviewResDto 간 변환 기능 구현
- 리뷰 생성 및 조회 시 데이터 변환을 위해 ReviewMapper 활용
- 엔티티와 DTO를 분리하여 유지보수성을 향상

shushuaka committed:
feat: 리뷰 관련 HTTP 요청 처리 컨트롤러 추가
- 리뷰 조회, 생성, 수정, 삭제를 위한 API 엔드포인트 구현
- GET /api/reviews: 리뷰 목록 조회
- POST /api/reviews: 새로운 리뷰 생성
- PATCH /api/reviews/{reviewId}: 기존 리뷰 수정
- DELETE /api/reviews/{reviewId}: 특정 리뷰 삭제
- 유스케이스를 호출하여 비즈니스 로직과 HTTP 요청 처리 분리

shushuaka committed:
feat: 리뷰 관련 HTTP 요청 처리 컨트롤러 추가
- 리뷰 조회, 생성, 수정, 삭제를 위한 API 엔드포인트 구현
- GET /api/reviews: 리뷰 목록 조회
- POST /api/reviews: 새로운 리뷰 생성
- PATCH /api/reviews/{reviewId}: 기존 리뷰 수정
- DELETE /api/reviews/{reviewId}: 특정 리뷰 삭제
- 유스케이스를 호출하여 비즈니스 로직과 HTTP 요청 처리 분리

shushuaka committed:
feat: 리뷰 관련 예외 처리 로직 추가
- ExceptionResponseStatus에 리뷰 관련 상태(REVIEW_NOT_FOUND, REVIEW_ALREADY_EXISTS, INVALID_REVIEW_REQUEST) 추가
- CustomException을 통해 도메인별 예외 처리 구조 구현
- ExceptionResponse를 활용하여 클라이언트로 명확한 응답 반환

shushuaka committed:
feat: 리뷰 관련 예외 처리 로직 추가
- ExceptionResponseStatus에 리뷰 관련 상태(REVIEW_NOT_FOUND, REVIEW_ALREADY_EXISTS, INVALID_REVIEW_REQUEST) 추가
- CustomException을 통해 도메인별 예외 처리 구조 구현
- ExceptionResponse를 활용하여 클라이언트로 명확한 응답 반환

shushuaka committed:
feat: 리뷰 관련 컨트롤러 추가
- 리뷰 조회, 생성, 수정, 삭제 API 구현
- 서비스 호출로 비즈니스 로직 처리

shushuaka committed:
feat: 리뷰 관련 컨트롤러 추가
- 리뷰 조회, 생성, 수정, 삭제 API 구현
- 서비스 호출로 비즈니스 로직 처리

shushuaka committed:
feat: 리뷰 서비스 추가
- 리뷰 생성, 조회, 수정, 삭제 비즈니스 로직 구현
- 예외 처리와 데이터 일관성 보장

shushuaka committed:
feat: 리뷰 서비스 추가
- 리뷰 생성, 조회, 수정, 삭제 비즈니스 로직 구현
- 예외 처리와 데이터 일관성 보장

shushuaka committed:
Merge remote-tracking branch 'origin/feat/review' into feat/review

# Conflicts:
#	src/main/kotlin/com/sparta/interparty/global/exception/ExceptionResponseStatus.kt

shushuaka commented:
아 커밋할 때 뒤에 #1 붙였어야했는데, 정신머리가 없었네요ㅜㅜ 죄송합니다...

coldrice99 commented on src/main/kotlin/com/sparta/interparty/domain/review/dto/req/ReviewReqDto.kt:
```diff
@@ -1,4 +1,16 @@
+
+    @field:Min(value = 1, message = "평점은 최소 1점 이상이어야 합니다.")
+    @field:Max(value = 10, message = "평점은 최대 10점까지 입력 가능합니다.")
+    val rating: Int
```
validation 예외 메세지 잘 나오는지 테스트 해보셨나요?!
```
// 유효성 검사 예외의 처리
    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleException(ex: MethodArgumentNotValidException): ResponseEntity<ExceptionResponse> {
        val msg: String = ex.bindingResult.allErrors[0].defaultMessage?: "입력 값이 유효하지 않습니다."
        return ExceptionResponseStatus.BAD_REQUEST.toResponseEntity(msg)
    }
```
   글로벌 헨들러의 이 부분에서 validation 예외가 "입력 값이 유효하지 않습니다."로 통일 처리 되는 것으로 알고 있습니다. 근데 자꾸 서버 에러(500)로 걸려서 아직 제대로 안되긴 합니다만 궁금해서 여쭤봐요!

34-43 replied:
그거 제가 따로 이슈 팠으니 일단 스킵~~

coldrice99 commented on src/main/kotlin/com/sparta/interparty/domain/review/entity/Review.kt:
```diff
@@ -1,4 +1,38 @@
+    fun update(comment: String, rating: Int) {
+        this.comment = comment
+        this.rating = rating
+        this.updatedAt = LocalDateTime.now()
```
timestamp 상속 하시고 createdAt, updatedAt은 지워주시면 더 좋을 거 같아요!

34-43 replied:
TMI:   timestamped에 updatedAt 대신 modifiedAt 으로 되어있는데, 이유는 불명.. 일단 후자로 통일해서 쓰시죠

34-43 replied:
그리고 지금 dev에 있는 timestamped는 직접 참조 안되고 getter 써야합니다

coldrice99 commented on src/main/kotlin/com/sparta/interparty/global/exception/CustomException.kt:
```diff
@@ -1,4 +1,8 @@
+    RuntimeException(exceptionResponseStatus.message)
+
+// auth
+class DuplicateUsernameException(val exceptionResponseStatus: ExceptionResponseStatus) :
```
이 부분은 제가 지웠었는데 dev에서 최신화 한번 해주시면 감사하겠습니다ㅏ

34-43 commented on src/main/kotlin/com/sparta/interparty/global/exception/ExceptionResponseStatus.kt:
```diff
@@ -19,11 +19,15 @@ enum class ExceptionResponseStatus(val status: HttpStatus, val message: String)
     // user
-    DELETED_USER(HttpStatus.FORBIDDEN,"이미 탈퇴한 사용자입니다.")
+    DELETED_USER(HttpStatus.FORBIDDEN,"이미 탈퇴한 사용자입니다.");
     ;
```
31번행에 세미콜론 달아뒀으니 위에 있는 Case 들은 확장성 고려해서 쉼표 붙여주시면 됩니다~

shushuaka replied:
근데 저 두개는 중복되는거라서 삭제 했는데요...

34-43 commented on src/main/kotlin/com/sparta/interparty/global/exception/ExceptionResponse.kt:
```diff
@@ -6,6 +6,8 @@ data class ExceptionResponse(
+    constructor(exceptionResponseStatus: ExceptionResponseStatus)
+            : this(exceptionResponseStatus.status.value(), exceptionResponseStatus.message)
+    constructor(status: HttpStatus, message: String)
+            : this(status.value(), message)
```
코드 변경이 없는 것 같은데 책임 유지하기 위해서 롤백 부탁드리겠습니다

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/controller/ReviewController.kt:
```diff
@@ -1,4 +1,45 @@
+        @RequestBody dto: ReviewReqDto
+    ): ResponseEntity<ReviewResDto> {
+        val review = reviewService.createReview(userId, showId, dto)
+        return ResponseEntity.ok(review)
```
나머지 domain은 Created(201)로 통일했으니
`ResponseEntity.status(HttpStatus.CREATED).body(review)`
로 수정 부탁드립니다~

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/controller/ReviewController.kt:
```diff
@@ -1,4 +1,45 @@
+
+    @GetMapping
+    fun getReviews(@RequestParam(required = false) showId: Long?): ResponseEntity<List<ReviewResDto>> {
+        val reviews = reviewService.getReviews(showId)
```
여기도 레퍼런스 생각하면 페이징 넣는 게 적합해 보이는데, 귀찮으니까 패스하는거에 동의합니다

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/controller/ReviewController.kt:
```diff
@@ -1,4 +1,45 @@
+        @PathVariable reviewId: Long,
+        @RequestBody dto: ReviewReqDto
+    ): ResponseEntity<ReviewResDto> {
+        val updatedReview = reviewService.updateReview(reviewId, dto)
```
서비스 로직으로 dto 종속되는 구조 발견

shushuaka replied:
DTO로 책임 분리 완료!

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/controller/ReviewController.kt:
```diff
@@ -1,4 +1,45 @@
+    }
+
+    @DeleteMapping("/{reviewId}")
+    fun deleteReview(@PathVariable reviewId: Long): ResponseEntity<Void> {
```
Void -> Unit

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/controller/ReviewController.kt:
```diff
@@ -1,4 +1,45 @@
+) {
+
+    @GetMapping
+    fun getReviews(@RequestParam(required = false) showId: Long?): ResponseEntity<List<ReviewResDto>> {
```
Show.id 타입 = UUID

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/dto/res/ReviewResDto.kt:
```diff
@@ -1,4 +1,13 @@
+import java.time.LocalDateTime
+
+data class ReviewResDto(
+    val id: Long,
```
Long -> UUID

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/entity/Review.kt:
```diff
@@ -1,4 +1,38 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    val id: Long = 0,
```
@GeneratedValue(stratege = GenerationType.UUID)
var id: UUID? = null

nullable이나 기본값은 원하시는대로 하시되, 타입은 UUID로 변경 필요

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/entity/Review.kt:
```diff
@@ -1,4 +1,38 @@
+    val userId: Long,
+
+    @Column(nullable = false)
+    val showId: Long,
```
userId showId 둘 다 UUID

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/mapper/ReviewMapper.kt:
```diff
@@ -0,0 +1,29 @@
+        return ReviewResDto(
+            id = entity.id,
+            userId = entity.userId,
+            showId = entity.showId,
```
연관을 분리하셔서 문제점이 되지는 않겠지만, 나중에 연관이 필요해졌을 때는 showId가 꼭 들어갈 필요는 없겠네요 여기

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/repo/ReviewRepository.kt:
```diff
@@ -1,4 +1,15 @@
+
+    // 특정 쇼에 포함된 모든 리뷰를 조회
+    fun findAllByShowId(showId: Long): List<Review>
+}
```
상속 제네릭 포함해서 전부 UUID로 변경

34-43 replied:
SoftDelete 구현 후에, 조회 로직에 해당 필드 필터링 추가 필요

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/sevice/ReviewService.kt:
```diff
@@ -1,4 +1,55 @@
+    fun deleteReview(reviewId: Long) {
+        val review = reviewRepository.findById(reviewId)
+            .orElseThrow { CustomException(ExceptionResponseStatus.REVIEW_NOT_FOUND) }
+        reviewRepository.delete(review)
```
SoftDelete 구현하기로 하였으니, Dirty Checking 활용해서 필드 갱신 (후 save 호출) 으로 변경 필요

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/sevice/ReviewService.kt:
```diff
@@ -1,4 +1,55 @@
+    }
+
+    @Transactional
+    fun updateReview(reviewId: Long, requestDto: ReviewReqDto): ReviewResDto {
```
서비스에서 Dto에 의존하고 있는 로직
필드 2개밖에 안되니까 둘 다 그냥 매개변수로 받으면 될 것 같아요

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/usecase/CreateReviewUseCase.kt:
```diff
@@ -0,0 +1,25 @@
+        val existingReview = reviewRepository.findByUserIdAndShowId(userId, showId)
+        if (existingReview != null) {
+            throw CustomException(ExceptionResponseStatus.REVIEW_ALREADY_EXISTS)
+        }
```
existingReview를 Optional이 아니라 nullable로 받으면 17~19행을 한 줄로 변경 가능

`existingReview ?: throw CustomException(ExceptionResponseStatus.REVIEW_ALREADY_EXISTS)`

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/usecase/UpdateReviewUseCase.kt:
```diff
@@ -0,0 +1,26 @@
+        val review = reviewRepository.findById(reviewId)
+            .orElseThrow { CustomException(ExceptionResponseStatus.REVIEW_NOT_FOUND) }
+
+        if (dto.comment.isBlank() || dto.rating !in 1..10) {
```
오 좋은 로직이네요
dto validation에서 @field:Pattern 이런것으로도 적용 가능할듯

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/usecase/GetReviewsUseCase.kt:
```diff
@@ -0,0 +1,27 @@
+        }
+
+        if (reviews.isEmpty()) {
+            throw CustomException(ExceptionResponseStatus.REVIEW_NOT_FOUND)
```
리뷰가 없는 게 Spring 앱과 서비스 측면에서 장애나 예외 상황은 아니니까 장애 코드랑 예외를 굳이 던질 필요는 없는 것 같다고 생각합니다~

34-43 commented on src/main/kotlin/com/sparta/interparty/domain/review/usecase/GetReviewsUseCase.kt:
```diff
@@ -0,0 +1,27 @@
+        val reviews = if (showId != null) {
+            reviewRepository.findAllByShowId(showId)
+        } else {
+            reviewRepository.findAll()
```
showId가 null일 때를 로직 트리거로 사용하는 게 조금 비직관적일 수 있다고 생각됩니다
(+showId를 인자로 받았지만, 여러 문제로 누락되었을 때 예상치 못한 로직이 발생하는 문제가 있음)
오버로딩으로 해결하거나, 최소한 주석 한 줄 다는 정도면 충분할 것 같습니다!

shushuaka committed:
fix: 재혁님 의견 적극 반영

shushuaka committed:
Merge remote-tracking branch 'origin/dev' into feat/review

# Conflicts:
#	src/main/kotlin/com/sparta/interparty/domain/show/repo/ShowRepository.kt
#	src/main/kotlin/com/sparta/interparty/global/exception/ExceptionResponse.kt
#	src/main/kotlin/com/sparta/interparty/global/exception/ExceptionResponseStatus.kt

shushuaka committed:
fix: 예외 중복 정리

shushuaka committed:
feat: JWT 처리용 JwtUtil 클래스 구현 및 설정 추가

- JWT 생성, 검증, Claims 추출 등을 관리하는 JwtUtil 클래스 구현
- JWT를 쿠키에 저장하고 HTTP 요청에서 가져오는 기능 추가
- @Value를 사용해 Secret Key 초기화 및 인코딩/디코딩 오류 처리 로직 추가
- application.yml에 JWT Secret Key 설정 추가
- JWT 관련 예외 상황에 대한 로깅 개선

shushuaka commented:
와우~ perfect!! 🐧
